From a2594baaea3681508a4bd3a24e0407f2fecf8ba8 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Mon, 6 Nov 2023 14:23:29 +0100
Subject: [PATCH 0891/1010] evl/lib: add dynamically-growable cache

We have multiple upcoming uses of a fast cache in the oob netstack
(e.g. IPv4 routes, ARP entries) which must be accessed locklessly for
lookups from any execution stage, and may be updated under lock from
the in-band stage exclusively.

This patch introduces a generic cache supporting these capabilities,
with RCU-based read protection from updates, and lock-based
serialization for dealing with concurrent updates.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/evl/cache.h     | 111 +++++++++++
 kernel/evl/lib/Makefile |   2 +-
 kernel/evl/lib/cache.c  | 400 ++++++++++++++++++++++++++++++++++++++++
 3 files changed, 512 insertions(+), 1 deletion(-)
 create mode 100644 include/evl/cache.h
 create mode 100644 kernel/evl/lib/cache.c

diff --git a/include/evl/cache.h b/include/evl/cache.h
new file mode 100644
index 0000000000..846503eb64
--- /dev/null
+++ b/include/evl/cache.h
@@ -0,0 +1,111 @@
+/*
+ * SPDX-License-Identifier: GPL-2.0
+ *
+ * Copyright (C) 2023 Philippe Gerum  <rpm@xenomai.org>
+ *
+ * A dynamically-growable cache with support for in-band update and
+ * out-of-band lookup.
+ */
+
+#ifndef _EVL_CACHE_H
+#define _EVL_CACHE_H
+
+#include <linux/spinlock.h>
+#include <linux/rcupdate.h>
+#include <linux/refcount.h>
+#include <evl/work.h>
+
+struct evl_cache;
+
+/* Cacheable entry. */
+struct evl_cache_entry {
+	/* Out-of-band reference count. */
+	refcount_t refcnt;
+	/* Link in hash table. */
+	struct evl_cache_entry __rcu *next;
+	/* Holder for work trampoline to RCU. */
+	struct evl_work work;
+	/* Owner cache. */
+	struct evl_cache *cache;
+	/* RCU holder for release. */
+	struct rcu_head	rcu;
+};
+
+/* Client-specific operations. */
+struct evl_cache_ops {
+	u32 (*hash)(const void *key);
+	bool (*eq)(const struct evl_cache_entry *entry, const void *key);
+	void (*drop)(struct evl_cache_entry *entry);
+	const void *(*get_key)(const struct evl_cache_entry *entry);
+	char *(*format_key)(const struct evl_cache_entry *entry);
+};
+
+/* Dynamic hash table indexing entries. */
+struct evl_hash_table {
+	/* RCU-managed bucket array. */
+	struct evl_cache_entry __rcu **buckets;
+	/* Size order for bucket array. */
+	unsigned int shift;
+	/* Number of busy entries. */
+	int nr_entries;
+	/* RCU holder for release. */
+	struct rcu_head	rcu;
+};
+
+/* Generic cache. */
+struct evl_cache {
+	/* Client-specific operation descriptor. */
+	struct evl_cache_ops *ops;
+	/* For serializing updates (in-band only). */
+	spinlock_t lock;
+	/* Hash table of entries. */
+	struct evl_hash_table __rcu *hash_table;
+	/* Initial cache shift. */
+	size_t init_shift;
+	/* Name of cache. */
+	const char *name;
+};
+
+int evl_init_cache(struct evl_cache *cache);
+
+void evl_cleanup_cache(struct evl_cache *cache);
+
+int evl_add_cache_entry(struct evl_cache *cache,
+			struct evl_cache_entry *entry);
+
+int evl_add_cache_entry_locked(struct evl_cache *cache,
+			struct evl_cache_entry *entry);
+
+void evl_del_cache_entry(struct evl_cache *cache,
+			const void *key);
+
+void evl_del_cache_entry_locked(struct evl_cache *cache,
+			const void *key);
+
+struct evl_cache_entry *evl_lookup_cache(struct evl_cache *cache,
+					const void *key);
+
+static inline void evl_get_cache_entry(struct evl_cache_entry *entry)
+{
+	refcount_inc(&entry->refcnt);
+}
+
+void evl_put_cache_entry(struct evl_cache_entry *entry);
+
+void evl_clean_cache(struct evl_cache *cache,
+		bool (*testfn)(struct evl_cache_entry *e, void *arg),
+		void *arg);
+
+void evl_flush_cache(struct evl_cache *cache);
+
+static inline void evl_lock_cache(struct evl_cache *cache)
+{
+	spin_lock_bh(&cache->lock);
+}
+
+static inline void evl_unlock_cache(struct evl_cache *cache)
+{
+	spin_unlock_bh(&cache->lock);
+}
+
+#endif /* !_EVL_CACHE_H */
diff --git a/kernel/evl/lib/Makefile b/kernel/evl/lib/Makefile
index f49e5dc94c..d6bdc31113 100644
--- a/kernel/evl/lib/Makefile
+++ b/kernel/evl/lib/Makefile
@@ -1,3 +1,3 @@
 obj-y += lib.o
 
-lib-y := uio.o
+lib-y := cache.o uio.o
diff --git a/kernel/evl/lib/cache.c b/kernel/evl/lib/cache.c
new file mode 100644
index 0000000000..82c6faf13e
--- /dev/null
+++ b/kernel/evl/lib/cache.c
@@ -0,0 +1,400 @@
+/*
+ * SPDX-License-Identifier: GPL-2.0
+ *
+ * Copyright (C) 2023 Philippe Gerum  <rpm@xenomai.org>
+ *
+ * This file implements a generic dynamically-growable cache with
+ * support for in-band update and out-of-band lookup.
+ */
+
+#include <linux/kmemleak.h>
+#include <evl/cache.h>
+
+static int realloc_hash_table(struct evl_cache *cache);
+
+static void hash_free_rcu(struct rcu_head *rcu);
+
+static void entry_free_rcu(struct rcu_head *rcu);
+
+static void entry_free_work(struct evl_work *work);
+
+int evl_init_cache(struct evl_cache *cache) /* in-band */
+{
+	spin_lock_init(&cache->lock);
+	cache->hash_table = NULL;
+
+	return realloc_hash_table(cache);
+}
+EXPORT_SYMBOL_GPL(evl_init_cache);
+
+void evl_cleanup_cache(struct evl_cache *cache) /* in-band */
+{
+	evl_flush_cache(cache);
+}
+EXPORT_SYMBOL_GPL(evl_cleanup_cache);
+
+/*
+ * Cache a new entry. The cache must have been locked prior to calling
+ * this routine. This call must be issued from the in-band stage.
+ *
+ * @entry entry to add to the cache.
+ */
+int evl_add_cache_entry_locked(struct evl_cache *cache, /* in-band */
+			struct evl_cache_entry *entry)
+{
+	const void *key = cache->ops->get_key(entry);
+	struct evl_cache_entry *e, **ep;
+	struct evl_hash_table *ht;
+	u32 hashval;
+	int ret;
+
+	hashval = cache->ops->hash(key);
+retry:
+	ht = rcu_dereference_protected(cache->hash_table,
+				lockdep_is_held(&cache->lock));
+	if (!ht || ht->nr_entries >= (1 << ht->shift)) {
+		spin_unlock_bh(&cache->lock);
+		ret = realloc_hash_table(cache);
+		spin_lock_bh(&cache->lock);
+		if (ret)
+			return ret;
+		goto retry;
+	}
+
+	/*
+	 * Find the place to attach the new entry, dropping a previous
+	 * match on the fly if any.
+	 */
+	hashval >>= (32 - ht->shift); /* Modulo table size. */
+	for (ep = &ht->buckets[hashval],
+		     e = rcu_dereference_protected(ht->buckets[hashval],
+			     lockdep_is_held(&cache->lock));
+	     e; e = rcu_dereference_protected(e->next,
+		     lockdep_is_held(&cache->lock))) {
+		if (cache->ops->eq(e, key)) {
+			rcu_assign_pointer(*ep, e->next);
+			if (refcount_dec_and_test(&e->refcnt))
+				call_rcu(&e->rcu, entry_free_rcu);
+		} else {
+			ep = &e->next;
+		}
+	}
+
+	ht->nr_entries++;
+	entry->cache = cache;
+	refcount_set(&entry->refcnt, 1);
+	evl_init_work(&entry->work, entry_free_work);
+	rcu_assign_pointer(*ep, entry);
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(evl_add_cache_entry_locked);
+
+/*
+ * Cache a new entry. This call must be issued from the in-band
+ * stage.
+ *
+ * @entry entry to add to the cache.
+ */
+int evl_add_cache_entry(struct evl_cache *cache, /* in-band */
+			struct evl_cache_entry *entry)
+{
+	int ret;
+
+	spin_lock_bh(&cache->lock);
+	ret = evl_add_cache_entry_locked(cache, entry);
+	spin_unlock_bh(&cache->lock);
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(evl_add_cache_entry);
+
+/*
+ * Remove an entry from the cache.  The cache must have been locked
+ * prior to calling this routine. This call must be issued from the
+ * in-band stage.
+ *
+ * @entry entry to remove from the cache.
+ */
+void evl_del_cache_entry_locked(struct evl_cache *cache, const void *key) /* in-band */
+{
+	struct evl_cache_entry *e, **ep;
+	struct evl_hash_table *ht;
+	u32 hashval;
+
+	ht = rcu_dereference_protected(cache->hash_table,
+				lockdep_is_held(&cache->lock));
+	if (!ht)
+		return;
+
+	hashval = cache->ops->hash(key);
+	hashval >>= (32 - ht->shift); /* Modulo table size. */
+
+	/* Don't bark on invalid removal request, just ignore it. */
+	for (ep = &ht->buckets[hashval],
+		     e = rcu_dereference_protected(
+			     ht->buckets[hashval],
+			     lockdep_is_held(&cache->lock));
+	     e; e = rcu_dereference_protected(
+		     e->next,
+		     lockdep_is_held(&cache->lock))) {
+		if (cache->ops->eq(e, key)) {
+			rcu_assign_pointer(*ep, e->next);
+			if (refcount_dec_and_test(&e->refcnt))
+				call_rcu(&e->rcu, entry_free_rcu);
+			break;
+		}
+		ep = &e->next;
+	}
+}
+EXPORT_SYMBOL_GPL(evl_del_cache_entry_locked);
+
+/*
+ * Remove an entry from the cache. This call must be issued from the
+ * in-band stage.
+ *
+ * @entry entry to remove from the cache.
+ */
+void evl_del_cache_entry(struct evl_cache *cache, const void *key) /* in-band */
+{
+	spin_lock_bh(&cache->lock);
+	evl_del_cache_entry_locked(cache, key);
+	spin_unlock_bh(&cache->lock);
+}
+EXPORT_SYMBOL_GPL(evl_del_cache_entry);
+
+void evl_flush_cache(struct evl_cache *cache) /* in-band */
+{
+	struct evl_cache_entry __rcu **buckets, *e, *next;
+	struct evl_hash_table *ht;
+	size_t size;
+	int n;
+
+	spin_lock_bh(&cache->lock);
+
+	ht = rcu_dereference_protected(cache->hash_table,
+				lockdep_is_held(&cache->lock));
+	if (likely(ht)) {
+		rcu_assign_pointer(cache->hash_table, NULL);
+		size = (1 << ht->shift) * sizeof(struct evl_cache_entry *);
+		buckets = rcu_dereference_protected(ht->buckets,
+					lockdep_is_held(&cache->lock));
+
+		for (n = 0; n < (1 << ht->shift); n++) {
+			for (e = rcu_dereference_protected(buckets[n],
+					lockdep_is_held(&cache->lock)); e; e = next) {
+				next = rcu_dereference_protected(e->next,
+						lockdep_is_held(&cache->lock));
+				if (refcount_dec_and_test(&e->refcnt))
+					call_rcu(&e->rcu, entry_free_rcu);
+			}
+		}
+
+		call_rcu(&ht->rcu, hash_free_rcu);
+	}
+
+	spin_unlock_bh(&cache->lock);
+}
+EXPORT_SYMBOL_GPL(evl_flush_cache);
+
+void evl_clean_cache(struct evl_cache *cache,
+		bool (*testfn)(struct evl_cache_entry *e, void *arg),
+		void *arg) /* in-band */
+{
+	struct evl_cache_entry __rcu **buckets, *e, **ep, *next;
+	struct evl_hash_table *ht;
+	size_t size;
+	int n;
+
+	spin_lock_bh(&cache->lock);
+
+	ht = rcu_dereference_protected(cache->hash_table,
+				lockdep_is_held(&cache->lock));
+	if (likely(ht)) {
+		size = (1 << ht->shift) * sizeof(struct evl_cache_entry *);
+		buckets = rcu_dereference_protected(ht->buckets,
+					lockdep_is_held(&cache->lock));
+
+		for (n = 0; n < (1 << ht->shift); n++) {
+			for (ep = &ht->buckets[n],
+				     e = rcu_dereference_protected(buckets[n],
+					lockdep_is_held(&cache->lock)); e; e = next) {
+				next = rcu_dereference_protected(e->next,
+						lockdep_is_held(&cache->lock));
+				if (testfn(e, arg)) {
+					rcu_assign_pointer(*ep, e->next);
+					if (refcount_dec_and_test(&e->refcnt))
+						call_rcu(&e->rcu, entry_free_rcu);
+				} else {
+					ep = &e->next;
+				}
+			}
+		}
+	}
+
+	spin_unlock_bh(&cache->lock);
+}
+EXPORT_SYMBOL_GPL(evl_clean_cache);
+
+/* in-band / oob */
+struct evl_cache_entry *evl_lookup_cache(struct evl_cache *cache,
+					const void *key)
+{
+	struct evl_cache_entry *e = NULL;
+	struct evl_hash_table *ht;
+	u32 hashval;
+
+	rcu_read_lock();
+
+	ht = rcu_dereference(cache->hash_table);
+	if (likely(ht)) {
+		hashval = cache->ops->hash(key);
+		hashval >>= (32 - ht->shift); /* Modulo table size. */
+		for (e = rcu_dereference(ht->buckets[hashval]);
+		     e; e = rcu_dereference(e->next)) {
+			if (cache->ops->eq(e, key)) {
+				refcount_inc(&e->refcnt);
+				break;
+			}
+		}
+	}
+
+	rcu_read_unlock();
+
+	return e;
+}
+EXPORT_SYMBOL_GPL(evl_lookup_cache);
+
+/* in-band / oob */
+void evl_put_cache_entry(struct evl_cache_entry *entry)
+{
+	if (refcount_dec_and_test(&entry->refcnt)) {
+		if (running_inband())
+			call_rcu(&entry->rcu, entry_free_rcu);
+		else
+			evl_call_inband(&entry->work);
+	}
+}
+EXPORT_SYMBOL_GPL(evl_put_cache_entry);
+
+static void entry_free_rcu(struct rcu_head *rcu) /* in-band */
+{
+	struct evl_cache_entry *e = container_of(rcu, struct evl_cache_entry, rcu);
+	e->cache->ops->drop(e);
+}
+
+/*
+ * Trampoline to schedule our RCU callback. We could not do this
+ * directly from oob, since call_rcu() would not support this, _and_
+ * RCU does neither watch nor even know about the oob context in the
+ * first place.
+ */
+static void entry_free_work(struct evl_work *work) /* in-band */
+{
+	struct evl_cache_entry *e = container_of(work, struct evl_cache_entry, work);
+	call_rcu(&e->rcu, entry_free_rcu);
+}
+
+static void hash_free_rcu(struct rcu_head *head)
+{
+	struct evl_hash_table *ht =
+		container_of(head, struct evl_hash_table, rcu);
+	size_t size = (1 << ht->shift) * sizeof(struct evl_cache_entry *);
+	struct evl_cache_entry __rcu **buckets = ht->buckets;
+
+	if (size <= PAGE_SIZE) {
+		kfree(buckets);
+	} else {
+		kmemleak_free(buckets);
+		free_pages((unsigned long)buckets, get_order(size));
+	}
+
+	kfree(ht);
+}
+
+/*
+ * Grow the hash table of a cache.  We may look up into a table from
+ * oob inside a read-side RCU section (Dovetail may emulate an NMI
+ * entry in this case).
+ */
+static int realloc_hash_table(struct evl_cache *cache)
+{
+	struct evl_hash_table *old_ht, *new_ht;
+	struct evl_cache_entry __rcu **buckets;
+	unsigned int shift;
+	size_t size;
+	u32 hash;
+	int n;
+
+	rcu_read_lock();
+	old_ht = rcu_dereference(cache->hash_table);
+	shift = old_ht ? old_ht->shift + 1 : cache->init_shift;
+	rcu_read_unlock();
+	size = (1 << shift) * sizeof(struct evl_cache_entry *);
+
+	new_ht = kmalloc(sizeof(*new_ht), GFP_ATOMIC);
+	if (!new_ht)
+		return -ENOMEM;
+
+	if (size <= PAGE_SIZE) {
+		buckets = kzalloc(size, GFP_ATOMIC);
+	} else {
+		buckets = (struct evl_cache_entry __rcu **)
+			  __get_free_pages(GFP_ATOMIC | __GFP_ZERO,
+					   get_order(size));
+		kmemleak_alloc(buckets, size, 1, GFP_ATOMIC);
+	}
+
+	if (!buckets) {
+		kfree(new_ht);
+		return -ENOMEM;
+	}
+
+	/* The table must be valid before aborting. */
+	new_ht->buckets = buckets;
+	new_ht->shift = shift;
+	new_ht->nr_entries = 0;
+
+	spin_lock_bh(&cache->lock);
+
+	old_ht = rcu_dereference_protected(cache->hash_table, lockdep_is_held(&cache->lock));
+	if (!old_ht)
+		goto finish;
+
+	/* Somebody slipped in and grew the table already, abort. */
+	if (old_ht->shift >= shift) {
+		spin_unlock_bh(&cache->lock);
+		hash_free_rcu(&new_ht->rcu);
+		return 0;
+	}
+
+	/* Copy the old table content to the new (larger) one. */
+	for (n = 0; n < (1 << old_ht->shift); n++) {
+		struct evl_cache_entry *e, *next;
+
+		for (e = rcu_dereference_protected(old_ht->buckets[n],
+						   lockdep_is_held(&cache->lock));
+		     e; e = next) {
+			hash = cache->ops->hash(e);
+			hash >>= (32 - shift);
+			next = rcu_dereference_protected(e->next,
+						lockdep_is_held(&cache->lock));
+			rcu_assign_pointer(e->next,
+					   rcu_dereference_protected(
+						new_ht->buckets[hash],
+						lockdep_is_held(&cache->lock)));
+			rcu_assign_pointer(new_ht->buckets[hash], e);
+		}
+	}
+
+	new_ht->nr_entries = old_ht->nr_entries;
+
+finish:
+	rcu_assign_pointer(cache->hash_table, new_ht);
+	spin_unlock_bh(&cache->lock);
+
+	if (old_ht)
+		call_rcu(&old_ht->rcu, hash_free_rcu);
+
+	return 0;
+}
-- 
2.30.2

